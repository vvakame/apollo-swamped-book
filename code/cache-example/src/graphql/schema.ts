import { GraphQLSchema, GraphQLObjectType, GraphQLList } from "graphql";
import GraphQLJSON from "graphql-type-json";

const jsonData = [
  // { id: 1, name: "John Smith" },
  // { id: 2, name: "John Smith", data: { a: "a1" } },
  { id: 3, name: "John Smith", data: [{ a: "a1" }] },
  { id: 4, name: "Sara Smith", data: [{ a: "a2" }] },
  { id: 5, name: "Budd Deey", data: [{ a: "a3" }] }
];

const QueryType = new GraphQLObjectType({
  name: "Query",
  fields: {
    jsonData: {
      type: GraphQLJSON,
      resolve: (...args) => {
        console.log("jsonDataResolver", args);
        return jsonData[0];
      }
    },
    jsonDataList: {
      type: new GraphQLList(GraphQLJSON),
      resolve: () => jsonData
    }
  }
});

export const schema = new GraphQLSchema({ query: QueryType });
