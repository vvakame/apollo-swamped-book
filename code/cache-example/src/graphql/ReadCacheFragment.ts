/* tslint:disable */
/* eslint-disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL fragment: ReadCacheFragment
// ====================================================

export interface ReadCacheFragment {
  __typename: "Cat";
  id: string;
  kind: string | null;
  name: string;
}
