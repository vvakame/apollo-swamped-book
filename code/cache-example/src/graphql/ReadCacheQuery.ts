/* tslint:disable */
/* eslint-disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: ReadCacheQuery
// ====================================================

export interface ReadCacheQuery_cat {
  __typename: "Cat";
  id: string;
  name: string;
  kind: string | null;
}

export interface ReadCacheQuery {
  cat: ReadCacheQuery_cat | null;
}
