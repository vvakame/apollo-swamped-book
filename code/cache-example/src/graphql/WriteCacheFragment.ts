/* tslint:disable */
/* eslint-disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL fragment: WriteCacheFragment
// ====================================================

export interface WriteCacheFragment {
  __typename: "Cat";
  id: string;
  kind: string | null;
  name: string;
  data: any | null;
}
