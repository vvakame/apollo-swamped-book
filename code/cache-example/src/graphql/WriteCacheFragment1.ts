/* tslint:disable */
/* eslint-disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL fragment: WriteCacheFragment1
// ====================================================

export interface WriteCacheFragment1 {
  __typename: "Cat";
  id: string;
  kind: string | null;
}
