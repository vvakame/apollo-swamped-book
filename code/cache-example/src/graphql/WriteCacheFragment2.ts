/* tslint:disable */
/* eslint-disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL fragment: WriteCacheFragment2
// ====================================================

export interface WriteCacheFragment2 {
  __typename: "Cat";
  kind: string | null;
  name: string;
}
