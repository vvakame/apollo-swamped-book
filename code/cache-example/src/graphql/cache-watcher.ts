import { InMemoryCache, InMemoryCacheConfig } from 'apollo-cache-inmemory';
import { Cache } from 'apollo-cache';

export class CacheWatcher extends InMemoryCache {
  constructor(config: InMemoryCacheConfig) {
    super(config);
  }

  write(write: Cache.WriteOptions) {
    let before = Object.keys((this as any).data.data).length;
    super.write(write);
    let after = Object.keys((this as any).data.data).length;
    if (before === after) {
      console.warn(`cache.write is ignored`, write);
    }
  }
}

