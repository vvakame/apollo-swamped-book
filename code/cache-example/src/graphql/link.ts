import { graphql, print } from "graphql";
import { ApolloLink, Observable } from "apollo-link";
import { schema } from "./schema";

export const link = new ApolloLink(operation => {
  return new Observable(observer => {
    const { query, operationName, variables } = operation;

    (async () => {
      try {
        await delay(800);
        const result = await graphql(schema, print(query), null, null, variables, operationName);
        observer.next(result);
        observer.complete();
      } catch (e) {
        observer.error(e);
      }
    })();
  });
});

function delay(ms: number) {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve();
    }, ms);
  });
}
