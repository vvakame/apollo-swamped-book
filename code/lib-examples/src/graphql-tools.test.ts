import { makeExecutableSchema, addCatchUndefinedToSchema, addMockFunctionsToSchema } from "graphql-tools";
import { graphqlSync, graphql } from "graphql";

describe("graphql-tools", () => {
	// 全部抽出しているけど全部試すのはめんどくさい…
	// クライアント側開発に使えそうなものだけ取り上げる

	it("makeExecutableSchema", async () => {
		// 型定義とResolverを与えて実行可能なschemaを作る
		// #@@range_begin(makeExecutableSchema)
		// typeDefs に使える型はかなり幅広い
		// resolvers の型も色々
		const typeDefs = `
			type Query {
				foo: String
			}
		`;
		const resolvers = {
			Query: {
				foo: () => "foo!",
			},
		};

		const schema = makeExecutableSchema({
			typeDefs,
			resolvers,
		});

		const result = await graphql(schema, `{ foo }`);
		// #@@range_end(makeExecutableSchema)
		expect(result).toMatchSnapshot();
	});
	it("addCatchUndefinedToSchema", async () => {
		// resolverがundefinedを返したらエラーにしてくれる
		// ちゃんとnullを返しましょう
		const typeDefs = `
			type Query {
				foo: String
			}
		`;
		const resolvers = {
			Query: {
				foo: () => void 0,
			},
		};

		const schema = makeExecutableSchema({
			typeDefs,
			resolvers,
		});
		addCatchUndefinedToSchema(schema);

		const result = await graphql(schema, `{ foo }`);
		expect(result).toMatchSnapshot();
	});
	it.skip("addErrorLoggingToSchema", () => {
	});

	it.skip("mockServer", () => {
		// 単なる addMockFunctionsToSchema のラッパ
	});
	it("addMockFunctionsToSchema", async () => {
		// #@@range_begin(addMockFunctionsToSchema)
		// schemaのresolversに適当な値を詰めて返してくれるモック関数をセットしてくれる
		const typeDefs = `
			type Query {
				foo: String
				bar: Int
			}
		`;
		const resolvers = {
			Query: {
				foo: () => "foo!",
			},
		};

		const schema = makeExecutableSchema({
			typeDefs,
			resolvers,
		});
		addMockFunctionsToSchema({
			schema,
			mocks: {
				Int: () => 42,
			},
			preserveResolvers: true,
		});

		const result = await graphql(schema, `{ foo bar }`);
		// #@@range_end(addMockFunctionsToSchema)
		expect(result).toMatchSnapshot();
	});
	it.skip("MockList", () => {
	});

	it.skip("makeRemoteExecutableSchema", () => {
	});
	it.skip("introspectSchema", () => {
	});
	it.skip("mergeSchemas", () => {
	});
	it.skip("delegateToSchema", () => {
	});
	it.skip("defaultMergedResolver", () => {
	});
	it.skip("defaultCreateRemoteResolver", () => {
	});

	it.skip("transformSchema", () => {
	});
	it.skip("AddArgumentsAsVariables", () => {
	});
	it.skip("CheckResultAndHandleErrors", () => {
	});
	it.skip("ReplaceFieldWithFragment", () => {
	});
	it.skip("AddTypenameToAbstract", () => {
	});
	it.skip("FilterToSchema", () => {
	});
	it.skip("RenameTypes", () => {
	});
	it.skip("FilterTypes", () => {
	});
	it.skip("RenameTypes", () => {
	});
	it.skip("TransformRootFields", () => {
	});
	it.skip("RenameRootFields", () => {
	});
	it.skip("FilterRootFields", () => {
	});
	it.skip("ExpandAbstractTypes", () => {
	});
	it.skip("ExtractField", () => {
	});
	it.skip("WrapQuery", () => {
	});

	it.skip("SchemaDirectiveVisitor", () => {
	});
});
