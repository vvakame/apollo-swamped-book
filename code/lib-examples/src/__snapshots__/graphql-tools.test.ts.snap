// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql-tools addCatchUndefinedToSchema 1`] = `
Object {
  "data": Object {
    "foo": null,
  },
  "errors": Array [
    [GraphQLError: Resolve function for "Query.foo" returned undefined],
  ],
}
`;

exports[`graphql-tools addMockFunctionsToSchema 1`] = `
Object {
  "data": Object {
    "bar": 42,
    "foo": "foo!",
  },
}
`;

exports[`graphql-tools makeExecutableSchema 1`] = `
Object {
  "data": Object {
    "foo": "foo!",
  },
}
`;
